code/datums/components/echolocation.dm

```diff
@@ -0,0 +1,177 @@
+				var/datum/outputs/echo_override/O = D
+				image_output = mutable_appearance(O.vfx.icon, O.vfx.icon_state, O.vfx.layer, O.vfx.plane)
+				realign_icon(image_output, S)
+		show_image(receivers, image_output, S)
+	for(var/T in turfs)
+		var/key = generate_wall_key(T)
+		if(echo_images[key])
+			image_output = mutable_appearance(echo_images[key].icon, echo_images[key].icon_state, echo_images[key].layer, echo_images[key].plane)
+			image_output.filters = echo_images[key].filters
+		else
+			image_output = generate_wall_image(T)
+			image_output.loc = T
+			echo_images[generate_wall_key(T)] = image_output
+		show_image(receivers, image_output, T)
+
+/datum/component/echolocation/proc/show_image(list/receivers, mutable_appearance/mutable_echo, atom/input)
+	var/image/image_echo = image(mutable_echo)
+	if(istype(input, /turf))
+		image_echo.loc = input
+	else
+		image_echo.loc = input.loc
+	for(var/M in receivers)
+		var/mob/receiving_mob = M
+		if(receiving_mob.client)
+			receiving_mob.client.images += image_echo
+			addtimer(CALLBACK(src, .proc/remove_image, image_echo, receiving_mob), 30)
+
+/datum/component/echolocation/proc/remove_image(sound_image, mob/M)
+	if(M.client && sound_image)
+		M.client.images -= sound_image
+		qdel(sound_image)
+
+/datum/component/echolocation/proc/generate_image(atom/input)
+	var/icon/I
+	if(uniques[input.type])
+		I = getFlatIcon(input)
+	else
+		I = icon(input.icon, input.icon_state)
+	I.MapColors(rgb(0,0,0,0), rgb(0,0,0,0), rgb(0,0,0,255), rgb(0,0,0,-254))
+	var/mutable_appearance/final_image = mutable_appearance(I, input.icon_state, CURSE_LAYER, 50)
+	final_image.filters += filter(type="outline", size=1, color="#FFFFFF")
+	return final_image
+
+/datum/component/echolocation/proc/generate_wall_image(turf/input)
+	var/icon/I = icon('icons/obj/echo_override.dmi',"wall")
+	var/list/dirs = list()
+	for(var/direction in GLOB.cardinals)
+		var/turf/T = get_step(input, direction)
+		if(istype(T, /turf/closed))
+			dirs += direction
+	for(var/dir in dirs)
+		switch(dir)
+			if(NORTH)
+				I.DrawBox(null,2,32,31,31)
+			if(SOUTH)
+				I.DrawBox(null,2,1,31,1)
+			if(EAST)
+				I.DrawBox(null,32,2,32,31)
+			if(WEST)
+				I.DrawBox(null,1,2,1,31)
+	return mutable_appearance(I, null, CURSE_LAYER, 50)
+
+/datum/component/echolocation/proc/generate_wall_key(turf/input)
+	var/list/dirs = list()
+	for(var/direction in GLOB.cardinals)
+		var/turf/T = get_step(input, direction)
+		if(istype(T, /turf/closed))
+			dirs += direction
+	var/key = dirs.Join()
+	return key
+
+/datum/component/echolocation/proc/realign_icon(mutable_appearance/I, atom/input)
+	I.dir = input.dir
+	I.loc = input.loc
+	I.pixel_x = input.pixel_x
+	I.pixel_y = input.pixel_y
+
+//actions
+
+/datum/action/innate/echo
+	name = "Echolocate"
+	check_flags = AB_CHECK_CONSCIOUS
+	icon_icon = 'icons/mob/actions/actions_items.dmi'
+	button_icon_state = "meson"
+
+/datum/action/innate/echo/Activate()
+	SEND_SIGNAL(owner, COMSIG_ECHOLOCATION_PING)
```

Comment:
Avoid having signals with multiple sources or signals which are described by what they do/activate. Signals should be agnostic of the functionality that uses them.

What you can do instead is make use of the generic trigger action signal and have an action empty of functionality you use for it, then handle the logic in the component itself